<div dir = "rtl" style="direction:rtl;line-height:300%;">
    <font face="XB Zar" size=6 color=#F32500>
        <div  align=center>
            تمرین هفته دوازدهم بوت کمپ یادگیری ماشین-
            مپصا
        </div>
        <br/>
    </font>
</div>

In [1]:
# import all packages and libraries here.
import numpy as np
import pandas as pd

# 1- State with reasons whether the following sentences are true or false?

<ul style="font-size:25px">
    <li>
        SVMs are not suitable for large datasets.
    </li>
    <li>
       SVMs perform poorly in imbalanced datasets.
    </li>
     <li>
        SVMs perform poorly when there is just too much noise in the data.
    </li>
</ul>


<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=5 color=#f7f3c8>   <ul>
<li>
جواب:
<li>
            <li>
          	درست است. اگرچه SVM در مجموعه داده‌هایی با ابعاد متوسط به خوبی کار می‌کند، اما با افزایش تعداد ویژگی‌ها و نمونه‌ها، زمان آموزش مدل به شدت افزایش می‌یابد. این مسئله باعث می‌شود که استفاده از SVM در داده‌های بزرگ و پیچیده به لحاظ محاسباتی ناکارآمد باشد.
            </li>
            <li>
          	درست است. زمانی که داده‌های آموزشی نامتعادل هستند، یعنی تعداد نمونه‌ها در یک کلاس به مراتب بیشتر از کلاس‌های دیگر است، SVM ممکن است به سمت کلاس با تعداد نمونه بیشتر سوگیری پیدا کند. این مسئله می‌تواند منجر به تشخیص نادرست کلاس‌های با تعداد نمونه کمتر شود.	
            </li>
            <li>
           	اشتباه است. SVM بیش از حد روی نمونه‌های نویزدار تمرکز نمی‌کنند و حساسیت کمتری نسبت به موارد دورافتاده دارد، زیرا سعی می‌کند به‌جای حداقل کردن خطای طبقه‌بندی، حاشیه بین دو کلاس را به حداکثر برساند.
            </li>           
         </ul>
</div>

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=6>
		 ۲- دیتاست lsvt-voice-rehabilitation را از لینک زیر دانلود کنید و به سوالات زیر حواب دهید. در این تمرین مجاز به استفاده از کتابخانه scikit-learn می باشید.
		</br>
</div>
		https://archive.ics.uci.edu/ml/datasets/LSVT+Voice+Rehabilitation

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=6>
		 الف- داده ها را به روشهای زیر دسته بندی کنید (حواستان باشد که داده ها را قبل از اعمال به مدل، نرمالایز کنید).
        <ul>
            <li>
            کرنل خطی
            </li>
            <li>
            کرنل چندجمله ای (پارامترهای r, d)
            </li>
            <li>
            کرنل rbf - پارامتر گاما
            </li>
            <li>
            سیگمویید - پارامتر r
            </li>           
         </ul>
</div

# Read Data

In [11]:
import pip
pip.main(["install", "openpyxl"])

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.


Collecting openpyxl

Downloading openpyxl-3.1.2-py2.py3-none-any.whl (249 kB)

C:\Users\Radin\Desktop\ML-Basic\ML_Basic_venv\Lib\site-packages\pip\_vendor\rich\live.py:231: UserWarning: install 
"ipywidgets" for Jupyter support
  warnings.warn('install "ipywidgets" for Jupyter support')

Collecting et-xmlfile (from openpyxl)

Downloading et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)

Installing collected packages: et-xmlfile, openpyxl

Successfully installed et-xmlfile-1.1.0 openpyxl-3.1.2

[notice] A new release of pip is available: 23.3 -> 23.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip

0

In [229]:
xlsx = pd.ExcelFile('LSVT_voice_rehabilitation.xlsx')
X = pd.read_excel(xlsx, 'Data')
y = pd.read_excel(xlsx, 'Binary response')
y= np.ravel(y)

In [400]:
print('Features Info:\n', X.info())
print('Label Info:', y.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Columns: 310 entries, Jitter->F0_abs_dif to det_TKEO_std4_10_coef
dtypes: float64(309), int64(1)
memory usage: 305.3 KB
Features Info:
 None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Data columns (total 1 columns):
 #   Column                                     Non-Null Count  Dtype
---  ------                                     --------------  -----
 0   Binary class 1=acceptable, 2=unacceptable  126 non-null    int64
dtypes: int64(1)
memory usage: 1.1 KB
Label Info: None


# Import necessary libraries

In [230]:
# TODO
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import LinearSVC
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, f1_score
from sklearn.model_selection import train_test_split

# Split train and test sets

In [231]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)

# Scaling the features

In [232]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=6>
		 ب- معیار دقت و f1 را برای هریک از دسته بندی های قسمت الف به دست آورید. (برای هر یک از پارامترهای گفته شده حداقل سه مقدار مختلف در نظر بگیرید)
		</br>
</div>

# Define indexes calculation function

In [233]:
# TODO
def get_parameter(C_list, gamma_list, d_list, r_list):
    accuracy_linear_model = []
    accuracy_poly_model = []
    accuracy_rbf_model = []
    accuracy_sigmoid_model = []
    f1_score_linear_model = []
    f1_score_poly_model = []
    f1_score_rbf_model = []
    f1_score_sigmoid_model = []
    parameters_list = []
    for sub_c in C_list:
        for sub_gamma in gamma_list:
            for sub_d in d_list:
                for sub_r in r_list:
                    parameters_list.append([sub_c, sub_gamma, sub_d, sub_r])
                    #############################################Linear kernel
                    svc_clf_linear = SVC(kernel="linear", C=sub_c)
                    svc_clf_linear.fit(X_train_scaled, y_train)
                    y_pred_linear=svc_clf_linear.predict(X_test_scaled)
                    accuracy_linear_model.append(accuracy_score(y_test, y_pred_linear))
                    f1_score_linear_model.append(f1_score(y_test, y_pred_linear))
                    #############################################Poly kernel
                    svc_clf_poly = SVC(kernel="poly", degree=sub_d, coef0=sub_r, C=sub_c)
                    svc_clf_poly.fit(X_train_scaled, y_train)
                    y_pred_poly=svc_clf_poly.predict(X_test_scaled)
                    accuracy_poly_model.append(accuracy_score(y_test, y_pred_poly)) 
                    f1_score_poly_model.append(f1_score(y_test, y_pred_poly))
                    #############################################RBF kernel
                    svc_clf_rbf = SVC(kernel="rbf", gamma=sub_gamma, C=sub_c)
                    svc_clf_rbf.fit(X_train_scaled, y_train)
                    y_pred_rbf=svc_clf_rbf.predict(X_test_scaled)
                    accuracy_rbf_model.append(accuracy_score(y_test, y_pred_rbf))
                    f1_score_rbf_model.append(f1_score(y_test, y_pred_rbf))
                    #############################################Sigmoid kernel
                    svc_clf_sigmoid = SVC(kernel="sigmoid", gamma=sub_gamma, coef0=sub_r, C=sub_c)
                    svc_clf_sigmoid.fit(X_train_scaled, y_train)
                    y_pred_sigmoid=svc_clf_sigmoid.predict(X_test_scaled)
                    accuracy_sigmoid_model.append(accuracy_score(y_test, y_pred_sigmoid))
                    f1_score_sigmoid_model.append(f1_score(y_test, y_pred_sigmoid))
    df = pd.DataFrame(data=[parameters_list, accuracy_linear_model, f1_score_linear_model, accuracy_poly_model, f1_score_poly_model, accuracy_rbf_model, f1_score_rbf_model, accuracy_sigmoid_model, f1_score_sigmoid_model]).T 
    df.columns = ['C, gamma, d, r', 'accuracy_linear', 'f1_score_linear', 'accuracy_poly', 'f1_score_poly', 'accuracy_rbf', 'f1_score_rbf', 'accuracy_sigmoid', 'f1_score_sigmoid']
    df.to_csv('accuracy_kernels_svm.csv', index=False)
    return df

# Call the indexes calculation function

In [234]:
get_parameter([1, 1e-1, 1e-2], [1e-1, 1, 1e1], [3, 4, 5], [1, 2, 3])

C, gamma, d, r accuracy_linear f1_score_linear accuracy_poly  \
0       [1, 0.1, 3, 1]        0.846154             0.8      0.923077   
1       [1, 0.1, 3, 2]        0.846154             0.8      0.846154   
2       [1, 0.1, 3, 3]        0.846154             0.8      0.846154   
3       [1, 0.1, 4, 1]        0.846154             0.8      0.923077   
4       [1, 0.1, 4, 2]        0.846154             0.8      0.769231   
..                 ...             ...             ...           ...   
76  [0.01, 10.0, 4, 2]        0.923077        0.888889      0.769231   
77  [0.01, 10.0, 4, 3]        0.923077        0.888889      0.769231   
78  [0.01, 10.0, 5, 1]        0.923077        0.888889      0.692308   
79  [0.01, 10.0, 5, 2]        0.923077        0.888889      0.769231   
80  [0.01, 10.0, 5, 3]        0.923077        0.888889      0.846154   

   f1_score_poly accuracy_rbf f1_score_rbf accuracy_sigmoid f1_score_sigmoid  
0       0.888889     0.692308          0.0         0.692308         0.333333  
1            0.8     0.692308          0.0         0.692308         0.333333  
2            0.8     0.692308          0.0         0.692308         0.333333  
3       0.888889     0.692308          0.0         0.692308         0.333333  
4       0.666667     0.692308          0.0         0.692308         0.333333  
..           ...          ...          ...              ...              ...  
76           0.4     0.692308          0.0         0.692308              0.0  
77      0.571429     0.692308          0.0         0.692308              0.0  
78           0.0     0.692308          0.0         0.692308              0.0  
79      0.571429     0.692308          0.0         0.692308              0.0  
80           0.8     0.692308          0.0         0.692308              0.0  

[81 rows x 9 columns]

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=6>
		 ج- تاثیر پارامترهای هر کرنل را بر کارآیی مدل ها تحلیل کنید.
		</br>
</div>

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=5 color=#f7f3c8>   <ul>
            <li>  
            <li>
                پارامتر C یک مقدار مثبت است که با افزایش آن قیدهای مسئله بهینه سازی بیشتر شده و اگر خیلی بزرگ باشد، مسئله شبیه به hard margin می شود. این یعنی خطای طبقه بندی باید صفر شود.
            </li>
            <li>
          	پارامتر سیگما دارای یک مقدار مثبت است. اینکه هر نقطه مرکزی چقدر از همسایگی خودش را پوشش دهد بسته به مقدار سیگمای کرنل دارد و اگر به درستی تعیین نشود، داده ها در فضای جدید خطی نخواهند بود.	
            </li>
            <li>
           	هر چقدر پارامتر گاما بزرگتر باشد، الگوریتم تلاش می کند برازش را دقیقا براساس مجموعه داده های تمرینی انجام دهد  و این امر موجب تعمیم یافتن خطا و وقوع مشکل بیش برازش شود. این یعنی فضای کمتری به هر کلاس می دهد و درنتیجه برای داده تست جایی نباشد. پس موجب می شود خطای بایاس کم و خطای واریانس زیاد شود.
            </li>           
         </ul>
</div>

<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=6>
		 د- کدام مدل عملکرد بهتری دارد؟ چرا؟
		</br>
</div>


<div id="bayes" style="direction:rtl;line-height:300%;">
	<font face="XB Zar" size=5 color=#f7f3c8>   <ul>
       <li>
      پارامترهایی که برای کرنل های این الگوریتم انتخاب شده اند، براساس حدس و خطا بوده است و مقدار آنها بهینه نیست. براساس نتایج حاصل در جدول بالا، مقدارهای accuracy و f1-score برای کرنل خطی بیشتر از کرنل های دیگر بوده است. از آنجایی‎که کرنل خطی برای داده هایی که فیچرهای بالایی دارند مناسب تر از کرنل های دیگر است. باتوجه با اینکه دیتاست مذکور دارای 310 فیچر است، این دلیل می تواند صادق باشد
            </li>           
         </ul>
</div>

# 3- Student Intervention System

<div style="margin-left: 10px;font-size:25px">a) Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, 'passed', will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.</div>

In [31]:
student_data = pd.read_csv("student_data.csv")
student_data.head(2)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home  teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home    other  ...   

  internet romantic  famrel  freetime  goout Dalc Walc health absences passed  
0       no       no       4         3      4    1    1      3        6     no  
1      yes       no       5         3      3    1    1      3        4     no  

[2 rows x 31 columns]

<div style="margin-left: 10px;font-size:25px">
  <p>  
b) Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
    </p>
    <ul>
        <li>The total number of students, n_students.</li>
        <li>The total number of features for each student, n_features.</li>
        <li>The number of those students who passed, n_passed.</li>
        <li>The number of those students who failed, n_failed.</li>
        <li>The graduation rate of the class, grad_rate, in percent (%).</li>
     </ul>

</div>

In [43]:
student_data.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'passed'],
      dtype='object')

In [39]:
student_data.isnull().any()

school        False
sex           False
age           False
address       False
famsize       False
Pstatus       False
Medu          False
Fedu          False
Mjob          False
Fjob          False
reason        False
guardian      False
traveltime    False
studytime     False
failures      False
schoolsup     False
famsup        False
paid          False
activities    False
nursery       False
higher        False
internet      False
romantic      False
famrel        False
freetime      False
goout         False
Dalc          False
Walc          False
health        False
absences      False
passed        False
dtype: bool

In [40]:
student_data.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences  
count  395.000000  
mean     5.708861  
std      8.003096  
min      0.000000  
25%      0.000000  
50%      4.000000  
75%      8.000000  
max     75.000000

In [36]:
student_data.nunique()

school         2
sex            2
age            8
address        2
famsize        2
Pstatus        2
Medu           5
Fedu           5
Mjob           5
Fjob           5
reason         4
guardian       3
traveltime     4
studytime      4
failures       4
schoolsup      2
famsup         2
paid           2
activities     2
nursery        2
higher         2
internet       2
romantic       2
famrel         5
freetime       5
goout          5
Dalc           5
Walc           5
health         5
absences      34
passed         2
dtype: int64

In [54]:
# TODO: Calculate number of students

n_students = student_data['age'].count()

# TODO: Calculate number of features
n_features = student_data.shape[1]-1

# TODO: Calculate passing students
n_passed = student_data.loc[student_data['passed']=='yes', 'passed'].count()

# TODO: Calculate failing students
n_failed = student_data.loc[student_data['passed']=='no', 'passed'].count()

# TODO: Calculate graduation rate
grad_rate = n_passed / n_students

# Print the results
print("Total number of students: {}".format(n_students))
print("Number of features: {}".format(n_features))
print("Number of students who passed: {}".format(n_passed))
print("Number of students who failed: {}".format(n_failed))
print("Graduation rate of the class: {:.2f}%".format(grad_rate))

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 0.67%


<div style="margin-left: 10px;font-size:25px">
    <p>c) Preparing the Data</p>
    <p>In this section, we will prepare the data for modeling, training and testing.</p>
    <p>Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.</p>
</div>




In [56]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print("Feature columns:\n{}".format(feature_cols))
print("\nTarget column: {}".format(target_col))

# Separate the data into feature data and target data (X_all and y_all, respectively)
X = student_data[feature_cols]
y = student_data[target_col]

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed


<div style="margin-left: 10px;font-size:25px">
    <p>d) Preprocess Feature Columns</p>
    <p>As you can see, there are several non-numeric columns that need to be converted! Many of them are simply yes/no, e.g. internet. These can be reasonably converted into 1/0 (binary) values.</p>
    <p>
Other columns, like Mjob and Fjob, have more than two values, and are known as categorical variables. The recommended way to handle such a column is to create as many columns as possible values (e.g. Fjob_teacher, Fjob_other, Fjob_services, etc.), and assign a 1 to one of them and 0 to all others.

These generated columns are sometimes called dummy variables, and we will use the pandas.get_dummies() function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.</p>
</div>


In [57]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.items():

        # If data type is non-numeric, replace all binary values with 1/0
        if col_data.dtype == object and len(col_data.unique()) == 2:
            print("*****binary*****")
            print("col name: ", col, end="----")
            col_data_unique = col_data.unique()
            print("unique values: ", col_data_unique)
            col_data = col_data.replace(col_data_unique, [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object and len(col_data.unique()) != 2:
            print("*****categorical*****")
            print("col name: ", col, end="----")
            col_data_unique = col_data.unique()
            print("unique values: ", col_data_unique)
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_preprocessed = preprocess_features(X)

*****binary*****
col name:  school----unique values:  ['GP' 'MS']
*****binary*****
col name:  sex----unique values:  ['F' 'M']
*****binary*****
col name:  address----unique values:  ['U' 'R']
*****binary*****
col name:  famsize----unique values:  ['GT3' 'LE3']
*****binary*****
col name:  Pstatus----unique values:  ['A' 'T']
*****categorical*****
col name:  Mjob----unique values:  ['at_home' 'health' 'other' 'services' 'teacher']
*****categorical*****
col name:  Fjob----unique values:  ['teacher' 'other' 'services' 'health' 'at_home']
*****categorical*****
col name:  reason----unique values:  ['course' 'other' 'home' 'reputation']
*****categorical*****
col name:  guardian----unique values:  ['mother' 'father' 'other']
*****binary*****
col name:  schoolsup----unique values:  ['yes' 'no']
*****binary*****
col name:  famsup----unique values:  ['no' 'yes']
*****binary*****
col name:  paid----unique values:  ['no' 'yes']
*****binary*****
col name:  activities----unique values:  ['no' 'yes']


# Convert the target to 0 & 1

In [133]:
y = y.replace(y.unique(), [1, 0])

# count the raw and Preprocessed features

In [224]:
print('Number of raw features: ', len(X.columns))

Number of raw features:  30


In [223]:
print( 'Number of preprocessed features: ', len(X_preprocessed.columns))

Number of preprocessed features:  43


<div style="margin-left: 10px;font-size:25px">
    <p>e) Training and Testing Data Split</p>
    <p>split the data (both features and corresponding labels) into training and test sets.(Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).)</p>
</div>

# Split train and set sets

In [142]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.24, random_state=10)

<div style="margin-left: 10px;font-size:25px">
<p>f) In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in scikit-learn. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to training data and measure the F1 score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F1 score on the training set, and F1 score on the testing set.</p>
<p>The following supervised learning models are currently available in scikit-learn that you may choose from:</p>
    <ul>
        <li>Gaussian Naive Bayes (GaussianNB)</li>
        <li>K-Nearest Neighbors (KNeighbors)</li>
        <li>Stochastic Gradient Descent (SGDC)</li>
        <li>Support Vector Machines (SVM)</li>
        <li>Logistic Regression</li>
     </ul>
</div>    

<div style="margin-left: 10px;font-size:25px">
<p></p>
<p>Answer:</p>
    <ul>
        <li>Note: this dataset is small and has large features. Also, it has low graduation rate of the class. It seems the GNN, KNN and SVM models are better than other two models. because of;</li>
        <li>The GNB is simle and will converge quicker than models like logistic Regression. Also, this model can peform better than more complex models when the dataset is small.</li>
        <li>The SVM model is appropiet when the dataset is small and has large featurs.</li>
        <li>SGDC model is appropriate for huge datasets, beacuase of it is unable all the data fetch together to the</li>
    ram.

# Import necessary libraries

In [121]:
from time import time
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score, accuracy_score

# Define indexes calculation function

In [213]:
# TODO
def get_model_train(clf, X_train, X_test, y_train, y_test):
# Scaling features 
 scaler = StandardScaler()
 X_train_scaled = scaler.fit_transform(X_train)
 X_test_scaled = scaler.fit_transform(X_test)
# Create an instance
 clf_model = clf()

# learning the model
 start_train_time = time()
 clf_model.fit(X_train_scaled, np.ravel(y_train))
 end_train_time = time()
 deltha_train = end_train_time - start_train_time
 y_predic_train = clf_model.predict(X_train_scaled)

# evaluation the model
 start_test_time = time()
 y_predic_test = clf_model.predict(X_test_scaled)
 end_test_time = time()
 delta_test = end_test_time - start_test_time

 return pd.DataFrame({'classifier_name' : clf, 'accuracy_train' : accuracy_score(y_train, y_predic_train),
                                      'f1_train' : f1_score(y_train, y_predic_train), 'run_time_train [S]' : deltha_train,
                                      'accuracy_test' : accuracy_score(y_test, y_predic_test),
                                      'f1_test' : f1_score(y_test, y_predic_test), 'run_time_test [S]' : delta_test}, index=[0])

# Call the indexes calculation function

In [221]:
for model in [LinearSVC, GaussianNB, KNeighborsClassifier]:
    print(get_model_train(model, X_train, X_test, np.ravel(y_train), np.ravel(y_test)))

C:\Users\Radin\Desktop\ML-Basic\ML_Basic_venv\Lib\site-packages\sklearn\svm\_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
C:\Users\Radin\Desktop\ML-Basic\ML_Basic_venv\Lib\site-packages\sklearn\svm\_base.py:1250: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


                            classifier_name  accuracy_train  f1_train  \
0  <class 'sklearn.svm._classes.LinearSVC'>        0.753333  0.574713   

   run_time_train [S]  accuracy_test   f1_test  run_time_test [S]  
0             0.12211       0.715789  0.470588                0.0  
                            classifier_name  accuracy_train  f1_train  \
0  <class 'sklearn.naive_bayes.GaussianNB'>            0.71  0.562814   

   run_time_train [S]  accuracy_test  f1_test  run_time_test [S]  
0            0.003002       0.705263     0.44                0.0  
                                     classifier_name  accuracy_train  \
0  <class 'sklearn.neighbors._classification.KNei...        0.766667   

   f1_train  run_time_train [S]  accuracy_test   f1_test  run_time_test [S]  
0  0.573171            0.001002       0.642105  0.291667           0.014013  


In [215]:
get_model_train(GaussianNB, X_train, X_test, np.ravel(y_train), np.ravel(y_test))

classifier_name  accuracy_train  f1_train  \
0  <class 'sklearn.naive_bayes.GaussianNB'>            0.71  0.562814   

   run_time_train [S]  accuracy_test  f1_test  run_time_test [S]  
0            0.002004       0.705263     0.44                0.0